### ISSUE TITLE
Prefer matching fragments within a given { origin }

### ISSUE THREAD

**Issue by boxxxie** [unpoly/unpoly#197]
### Bug description

discovered this when my forms were set to target an inner element, but instead the first matching element on the page was selected (this is a page with a list of forms)

```jinja
<div>
    <div class="auto-submit d-flex"></div>
    <form
      {% if id %}id="{{id}}"{% else %}id="form_{{name}}"{% endif %}
      up-transition="cross-fade"
      up-target=".auto-submit"
      up-method="{{up-method|default:PATCH}}"
      up-autosubmit
      up-delay="{{up-delay|default:1000}}"
      enctype="multipart/form-data"
      up-history="false"
      action="{{action}}">
        {% block component %}{% endblock %}
        <div class="auto-submit d-flex"> </div>
    </form>
</div>
```
response
```jinja
<div class="auto-submit d-flex">
    <img src="/assets/Font-Awesome/svgs/solid/check-circle.svg">
    <p>Profile Upadted !</p>
</div>
```

my workaround/fix
```javascript

var selector_closest_fixed = false;
(function fix_selector_closest(){
  if(!selector_closest_fixed){
    selector_closest_fixed = true
    up.Selector.prototype.closest = function(element) {
      var parentElement;
      let child_target = element.querySelector(this.unionSelector);
      if (child_target){
        return child_target
      }
      else if (this.matches(element)) {
        return element;
      } else if ((parentElement = element.parentElement)) {
        return this.closest(parentElement);
      }
    };
  }
})();
```

**Comment by adam12** [comment]
Can you confirm you're using Unpoly 2?

**Comment by boxxxie** [comment]
i'm using `unpoly ~2.0.1` (whatever is latest on CDN)
i tried many selectors for solutions, even `up-id`
i spent a lot of time on this issue, trying to figure out why my form post was changing html in some random place on my page, nowhere near the origin.
the way `selector closest` works either is not intuitive, incorrect, or maybe the function should be renamed.

`closest` really sounds like it should try to do just querySelector(selector), because the `closest` selection will be that. but closest is doing a recursive search of parents, so it's name should probably be something similar to what it's actually doing, like `search_parents` or something

since unpoly has the `up-content` att, i think i can do an example without needing a server. i'll post it soon.

**Comment by boxxxie** [comment]
https://glitch.com/edit/#!/ember-mousy-canvas

js has my fix, i can't get the `up-document/up-content` working (maybe another bug)

i added some style to show what unpoly is changing, js has my fix, remove the fix and see the behaviour difference.

**Comment by adam12** [comment]
I removed your Javascript changes and updated both `up-target` to this.

```
up-target="& .auto-submit"
```

I believe it achieves what you were looking for. Can you confirm? https://glitch.com/edit/#!/radial-attractive-equinox

**Comment by boxxxie** [comment]
yes, maybe this should be emphasised in the documentation. i tried to use :origin in my target, but it didn't work, the `&` is very unfamiliar to me, and i didn't attempt to use it as a selector.

seems that origin should be implied, though.

would be common that the selector you used would break things for people? for me it seems like the intuitive behaviour and `&` is redundant.

**Comment by adam12** [comment]
I did try `:origin .auto-submit` as a selector and it appears to work. The `&` selector was linked in the v1 `up-target` docs, but that's not the case for v2. If you think you can make the docs more expressive in this case, please submit a PR. To be honest, `&` is more familiar to me than `:origin`, as I have past experience with Sass.

> seems that origin should be implied, though.

I haven't ran into any issues here, so I can't really agree. Maybe Henning can provide an opinion here.

I'm going to close this for now since `up.fragment.closest` documents that it only looks at itself and then it's ancestors, never its children. If Henning deems this a bug we can re-open.

Please feel free to submit a PR for the `:origin` documentation if you can find a way to improve it.

**Comment by triskweline** [comment]
Both `:origin` and `&` are supported, as [documented here](https://unpoly.com/origin).

Even without `:origin` in the selector, Unpoly 2 tries to [take a given `{ origin }` into account](https://unpoly.com/fragment-placement#interaction-origin-is-considered) when matching fragments. Currently that mostly looks for a parent element around the origin.

I haven't had a use case where I needed to auto-match an origin child, but I guess we could support that.

The fix wouldn't go in `up.Selector` though, but here:

https://github.com/unpoly/unpoly/blob/e555c48d93e636f8a9c6643c38688cc237ee6853/lib/assets/javascripts/unpoly/fragment.coffee#L1033-L1038
https://github.com/unpoly/unpoly/blob/e555c48d93e636f8a9c6643c38688cc237ee6853/lib/assets/javascripts/unpoly/classes/fragment_finder.coffee
https://github.com/unpoly/unpoly/blob/e555c48d93e636f8a9c6643c38688cc237ee6853/spec_app/spec/javascripts/up/fragment_spec.js.coffee#L63-L87

**Comment by triskweline** [comment]
Re-opening this issue for a while in case someone wants to tackle that change.
